File: src/year2015/day02/parse.cc
#include "parse.h"

#include <algorithm>
#include <cstddef>
#include <stdexcept>
#include <string>

auto Year2015::Day02::parse(const std::string& input) -> Boxes {
  Boxes boxes = {};
  Box box = {0, 0, 0};
  int sideIndex = 0;
  for (const auto& chr : input) {
    switch (chr) {
      case '0':
      case '1':
      case '2':
      case '3':
      case '4':
      case '5':
      case '6':
      case '7':
      case '8':
      case '9':
        box[sideIndex] *= 10;
        box[sideIndex] += chr - '0';
        break;
      case 'x':
        if (++sideIndex == 3) {
          throw std::runtime_error(
              "A cuboid is supposed to have only 3 sides on line " +
              std::to_string(boxes.size() + 1));
        }
        break;
      case '\n':
        if (sideIndex == 2) {
          sideIndex = 0;
          std::sort(box.begin(), box.end());
          boxes.push_back(box);
          for (size_t boxLengthIndex = 0; boxLengthIndex != 3;
               boxLengthIndex++) {
            box[boxLengthIndex] = 0;
          }
        } else {
          throw std::runtime_error("Expected three sides, found only " +
                                   std::to_string(sideIndex + 1) +
                                   " when parsing input.txt on line " +
                                   std::to_string(boxes.size() + 1));
        }
        break;
      default:
        throw std::runtime_error("Unexpected character at line " +
                                 std::to_string(boxes.size() + 1) +
                                 " when parsing input.txt");
        break;
    }
  }

  // Deal with non-newline endings gracefully
  if (sideIndex == 0) {
  } else if (sideIndex == 2) {
    std::sort(box.begin(), box.end());
    boxes.push_back(box);
  } else {
    throw std::runtime_error(
        "Expected three sides, found only " + std::to_string(sideIndex + 1) +
        " when parsing input.txt on line " + std::to_string(boxes.size() + 1));
  }
  return boxes;
}

File: src/year2015/day02/parse.rs
pub type Box = [u32; 3];
pub type Boxes = Vec<Box>;

pub fn parse(input: String) -> Result<Boxes, String> {
  let mut boxes = Boxes::new();
  let mut current_box: Box = [0, 0, 0];
  let mut side_index = 0;
  for chr in input.chars() {
    match chr {
      '0'..='9' => {
        current_box[side_index] *= 10;
        current_box[side_index] += (chr as u32) - ('0' as u32);
      }
      'x' => {
        side_index += 1;
        if side_index == 3 {
          return Err(format!(
            "A cuboid is supposed to have only 3 sides on line {}",
            boxes.len() + 1
          ));
        }
      }
      '\n' => {
        if side_index == 2 {
          side_index = 0;
          current_box.sort();
          boxes.push(current_box);
          let mut index = 0;
          while index < current_box.len() {
            current_box[index] = 0;
            index += 1;
          }
        } else {
          return Err(format!(
            "Expected three sides, found only {} when parsing input.txt on line {}",
            side_index + 1,
            boxes.len() + 1
          ));
        }
      }
      _ => {
        return Err(format!(
          "Unexpected character at line {} when parsing input.txt",
          boxes.len() + 1
        ));
      }
    }
  }
  if side_index == 0 {
  } else if side_index == 2 {
    current_box.sort();
    boxes.push(current_box);
  } else {
    return Err(format!(
      "Expected three sides, found only {} when parsing input.txt on line {}",
      side_index + 1,
      boxes.len() + 1
    ));
  }
  return Ok(boxes);
}
